modules/javafx.graphics/src/main/java/com/sun/glass/ui/HeaderButtonOverlay.java

```diff
@@ -0,0 +1,709 @@
+        if (buttonVerticalAlignment.get() == ButtonVerticalAlignment.CENTER) {
+            if (left) {
+                double offset = getButtonOffsetY(button1Height);
+                button1X = snapPositionX(button1X + offset);
+                button2X = snapPositionX(button2X + offset);
+                button3X = snapPositionX(button3X + offset);
+                totalWidth = snapSizeX(totalWidth + offset * 2);
+            } else {
+                double offset = getButtonOffsetY(button3Height);
+                button1X = snapPositionX(button1X - offset);
+                button2X = snapPositionX(button2X - offset);
+                button3X = snapPositionX(button3X - offset);
+                totalWidth = snapSizeX(totalWidth + offset * 2);
+            }
+
+            totalHeight = snapSizeY(getEffectiveButtonHeight());
+        } else {
+            totalHeight = snapSizeY(Math.max(button1Height, Math.max(button2Height, button3Height)));
+        }
+
+        Dimension2D currentSize = left ? metrics.get().leftInset() : metrics.get().rightInset();
+
+        // Update the overlay metrics if they have changed.
+        if (currentSize.getWidth() != totalWidth || currentSize.getHeight() != totalHeight) {
+            var empty = new Dimension2D(0, 0);
+            var size = new Dimension2D(totalWidth, totalHeight);
+            HeaderButtonMetrics newMetrics = left
+                ? new HeaderButtonMetrics(size, empty, buttonDefaultHeight.get())
+                : new HeaderButtonMetrics(empty, size, buttonDefaultHeight.get());
+            metrics.set(newMetrics);
+        }
+
+        layoutInArea(button1, button1X, getButtonOffsetY(button1Height), button1Width, button1Height,
+                     BASELINE_OFFSET_SAME_AS_HEIGHT, Insets.EMPTY, true, true,
+                     HPos.LEFT, VPos.TOP, false);
+
+        layoutInArea(button2, button2X, getButtonOffsetY(button2Height), button2Width, button2Height,
+                     BASELINE_OFFSET_SAME_AS_HEIGHT, Insets.EMPTY, true, true,
+                     HPos.LEFT, VPos.TOP, false);
+
+        layoutInArea(button3, button3X, getButtonOffsetY(button3Height), button3Width, button3Height,
+                     BASELINE_OFFSET_SAME_AS_HEIGHT, Insets.EMPTY, true, true,
+                     HPos.LEFT, VPos.TOP, false);
+    }
+
+    @Override
+    public boolean usesMirroring() {
+        return false;
+    }
+
+    @Override
+    public List<CssMetaData<? extends Styleable, ?>> getCssMetaData() {
+        return METADATA;
+    }
+
+    private static double boundedWidth(Node node) {
+        return node.isManaged() ? boundedSize(node.minWidth(-1), node.prefWidth(-1), node.maxWidth(-1)) : 0;
+    }
+
+    private static double boundedHeight(Node node) {
+        return node.isManaged() ? boundedSize(node.minHeight(-1), node.prefHeight(-1), node.maxHeight(-1)) : 0;
+    }
+
+    private static double boundedSize(double min, double pref, double max) {
```

Comment:
minor:
it would be nice to move this to some common utilities class (there is already such a method in c.s.j.scene.control.skin.Utils class).

along with two previous methods.

Reply:
We can also consider exposing this as API (not with this PR, though). If it's useful for us, it might also be useful for application developers.

Reply:
my earlier PR was shut down under the rubric "we could do better" but "we" never did.